### ISSUE TITLE
Most of the story tests failing when used with storybook/addon-a11y@8.5.0

### ISSUE THREAD

**Issue by certainlyakey** [chanzuckerberg/axe-storybook-testing#103]
Hey, thanks for the great tool!

I noticed that when it is used with a latest @storybook/addon-a11y (8.5.0), most of the tests fail with this error: 

```
page.evaluate: Error: Axe is already running. Use `await axe.run()` to wait for the previous run to finish before starting a new run.
    at Cv (http://127.0.0.1:8000/assets/axe-BHni_kjN.js:18:55391)
    at Object.oT [as run] (http://127.0.0.1:8000/assets/axe-BHni_kjN.js:24:93604)
    at eval (eval at evaluate (:234:30), <anonymous>:17:27)
    at async <anonymous>:260:30
```

I discovered [this comment](https://github.com/storybookjs/storybook/issues/29805#issuecomment-2594439211). Considering that case, I found out when I go back to 8.4.7 for @storybook/addon-a11y (as well as all the other Storybook-related packages), the issue goes away. 

Not sure if this will lead somewhere, and there's no documented issue yet for @storybook/addon-a11y itself.

**Comment by ahuth** [comment]
Thanks for reporting! I haven't tried using React 19 or storybook 8.5 yet. I'll look into this tomorrow my time.

**Comment by booc0mtaco** [comment]
I believe I have found the issue:

* `axe-storybook-testing` will inject a version of `axe-core` into the page, living at `window.axe`
* with the latest version of the storybook packages, it will ALSO inject a version on top of the **same location** (somewhere?)
* after the first test run completes, things get out of sync, as some of the state used by the version we inject remains, but the object on `window` has been (partially?) overwritten (We are now in the bad state)

Will spend some time confirming and testing whether this is fully the case. Discovered this locally by opening a test on previous storybook packages and newest ones, navigating to the iframe view of the test, and checking for `window.axe` in the console in each.

**Comment by ahuth** [comment]
Good finds! Confirmed that updating storybook to 8.5.0 and including storybook-addon-a11y results in this issue.

Haven't found a good way to resolve it yet, though 😐

**Comment by booc0mtaco** [comment]
Ideally there would be a way to intercept/modify the stories being tested. But, since we run `storybook build`, then run `axe-storybook-testing`, the built tests already include the usage of `@storybook/addons-a11y`. It's very tricky and probably not possible to stop `window.axe` from being overwritten.

I did find a solution that we can document here for the time-being. We can use [`storybook build --test`](https://storybook.js.org/docs/api/main-config/main-config-build) to disable the addon (and maybe other things...).

* update the command to run `storybook build --test && axe-storybook # or similar`
* in your main.(js|ts) file, add in the config you'd like to disable. In our case `build: { test: { disabledAddons: ['@storybook/addon-a11y'] } }`
* re-run the script

It took me a second to understand the storybook documentation on this; you need to add `--test` which won't do anything until you ALSO add some changes to main.(js|ts)

@mpelham @certainlyakey given your projects, would you be able to try this approach?

**Comment by certainlyakey** [comment]
Sure, I will try it out soon. Thank you for the research! On 23. Jan 2025, at 18:22, Andrew Holloway ***@***.***> wrote:﻿
Ideally there would be a way to intercept/modify the stories being tested. But, since we run storybook build, then run axe-storybook-testing, the built tests already include the usage of @storybook/addons-a11y. It's very tricky and probably not possible to stop window.axe from being overwritten.
I did find a solution that we can document here for the time-being. We can use storybook build --test to disable the addon (and maybe other things...).

update the command to run storybook build --test && axe-storybook # or similar
in your main.(js|ts) file, add in the config you'd like to disable. In our case build: { test: { disabledAddons: ***@***.***/addon-a11y'] } }
re-run the script

It took me a second to understand the storybook documentation on this; you need to add --test which won't do anything until you ALSO add some changes to main.(js|ts)
@mpelham @certainlyakey given your projects, would you be able to try this approach?

**Comment by valentinpalkovic** [comment]
Hi folks,

The Storybook Core Maintaner here which broke your integration for Storybook 8.5 🙃👋 

Via query parameters, the a11y addon can be deactivated by appending the following information to the URL when visiting a particular story

```tsx
http://localhost:6006/<story>?globals=a11y.manual:!true
```

Then axe from addon-a11y will not run!

#

**Comment by booc0mtaco** [comment]
> Hi folks,
> 
> The Storybook Core Maintaner here which broke your integration for Storybook 8.5 🙃👋
> 
> Via query parameters, the a11y addon can be deactivated by appending the following information to the URL when visiting a particular story
> 
> http://localhost:6006/<story>?globals=a11y.manual:!true
> Then axe from addon-a11y will not run! Can you try this out?

Hi @valentinpalkovic, thanks for the note! I'll try this out; sounds like it should work, even with stories that were built with the addon enabled. It's nice to have a few paths forward 💯 .

Will update with results once I get another 2 seconds

**Comment by booc0mtaco** [comment]
Documented using `--test` in a new wiki page to unblock folks needing to complete an upgrade: https://github.com/chanzuckerberg/axe-storybook-testing/wiki/FAQ-&-Troubleshooting 

I have a local patch that can work without any config changes, using the tip from @valentinpalkovic (thanks again!). Will test that and work toward a PR including this change.

cc @ahuth

**Comment by certainlyakey** [comment]
@booc0mtaco sorry for the delay, the `--test` "hack" did work for me. Although i reckon it's better to update to the latest version of axe-storybook-testing anyway now.

**Comment by certainlyakey** [comment]
And the latest version also resolved it for me. So i guess the issue can be closed now.
